File: worker/url-reference/src/tests/response.rs
// Copyright © 2026 @sudocelestia. All rights reserved.
//
// PROPRIETARY AND CONFIDENTIAL SOURCE CODE.
//
// No licence, permission or authorisation is granted to use, copy, modify,
// compile, execute, distribute, publish, sublicense or otherwise exploit this
// file, except to the limited extent unavoidably permitted by applicable law
// or GitHub's Terms of Service.
//
// See the LICENSE file at the repository root for the complete terms.

use crate::response::{Response, write_to};
use std::io::{self, Write};

#[derive(Default)]
struct RecordingWriter {
    data: Vec<u8>,
    flushes: usize,
    fail_flush: bool,
}

impl Write for RecordingWriter {
    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
        self.data.extend_from_slice(data);
        Ok(data.len())
    }

    fn flush(&mut self) -> io::Result<()> {
        self.flushes += 1;
        if self.fail_flush {
            return Err(io::Error::other("flush"));
        }
        Ok(())
    }
}

#[test]
fn response_write_requires_flush() {
    let response = Response {
        protocol_version: 1,
        operation_id: "url-reference",
        operation_version: 1,
        attempt_id: "attempt",
        request_nonce: "nonce",
        worker_id: "celestia-url-reference",
        worker_version: "1",
        status: "failed",
        output_media_type: None,
        output_length: None,
        output_sha256: None,
        output: None,
        diagnostics: Vec::new(),
        duration_ns: 0,
    };
    let mut writer = RecordingWriter::default();
    assert_eq!(write_to(&response, &mut writer), Ok(()));
    assert_eq!(writer.flushes, 1);
    assert!(!writer.data.is_empty());
    writer.fail_flush = true;
    assert_eq!(write_to(&response, &mut writer), Err(()));
    assert_eq!(writer.flushes, 2);
}

File: internal/operation/urlreference/protocol/response.go
// Copyright © 2026 @sudocelestia. All rights reserved.
//
// PROPRIETARY AND CONFIDENTIAL SOURCE CODE.
//
// No licence, permission or authorisation is granted to use, copy, modify,
// compile, execute, distribute, publish, sublicense or otherwise exploit this
// file, except to the limited extent unavoidably permitted by applicable law
// or GitHub's Terms of Service.
//
// See the LICENSE file at the repository root for the complete terms.

package workerprotocol

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"unicode/utf8"
)

func decodeResponse(data []byte, correlation Correlation, exitCode int) (Response, error) {
	if len(data) == 0 || len(data) > MaxResponseBytes || !utf8.Valid(data) {
		return Response{}, protocolError("response bounds or UTF-8")
	}
	if err := validateJSONFrame(data); err != nil {
		return Response{}, err
	}
	var response Response
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&response); err != nil {
		return Response{}, protocolError("decode")
	}
	if err := validateFields(data, response.Status); err != nil {
		return Response{}, err
	}
	if err := validateResponse(response, correlation, exitCode); err != nil {
		return Response{}, err
	}
	return response, nil
}

func DecodeResponseForRequestCorrelation(
	data []byte,
	request Request,
	exitCode int,
) (Response, error) {
	if !validIdentity(request.AttemptID) ||
		!validIdentity(request.RequestNonce) ||
		request.AttemptID == request.RequestNonce ||
		request.InputMediaType != MediaType {
		return Response{}, protocolError("request correlation")
	}
	return decodeResponse(data, correlationFor(request), exitCode)
}

func validateResponse(response Response, correlation Correlation, exitCode int) error {
	if err := validateIdentity(response, correlation); err != nil {
		return err
	}
	if err := validateDiagnostics(response); err != nil {
		return err
	}
	return validateStatus(response, correlation, exitCode)
}

func validateIdentity(response Response, correlation Correlation) error {
	if response.ProtocolVersion != ProtocolVersion ||
		response.OperationID != OperationID ||
		response.OperationVersion != OperationVersion ||
		response.AttemptID != correlation.attemptID ||
		response.RequestNonce != correlation.nonce ||
		response.WorkerID != WorkerID ||
		response.WorkerVersion != WorkerVersion {
		return protocolError("identity or version mismatch")
	}
	if response.DurationNS < 0 || response.DurationNS > MaxDurationNS {
		return protocolError("duration")
	}
	return nil
}

func validateDiagnostics(response Response) error {
	if len(response.Diagnostics) > MaxDiagnostics {
		return protocolError("diagnostic count")
	}
	for _, diagnostic := range response.Diagnostics {
		if !diagnosticCode.MatchString(diagnostic.Code) ||
			!utf8.ValidString(diagnostic.Message) ||
			len(diagnostic.Message) > MaxMessageBytes {
			return protocolError("diagnostic")
		}
	}
	return nil
}

func validateStatus(response Response, correlation Correlation, exitCode int) error {
	switch response.Status {
	case Completed:
		if exitCode != 0 {
			return protocolError("completed exit")
		}
		return validateOutput(response, correlation)
	case Rejected:
		if exitCode != 2 || len(response.Diagnostics) == 0 {
			return protocolError("rejected response")
		}
	case Failed:
		if exitCode != 3 || len(response.Diagnostics) == 0 {
			return protocolError("failed response")
		}
	default:
		return protocolError("status")
	}
	return nil
}

func validateOutput(response Response, correlation Correlation) error {
	if *response.OutputMediaType != correlation.mediaType ||
		!utf8.ValidString(*response.Output) ||
		*response.OutputLength < 1 ||
		*response.OutputLength > MaxOutputBytes ||
		len(*response.Output) != *response.OutputLength {
		return protocolError("output metadata")
	}
	hash := sha256.Sum256([]byte(*response.Output))
	if !lowerHex.MatchString(*response.OutputSHA256) ||
		*response.OutputSHA256 != hex.EncodeToString(hash[:]) {
		return protocolError("output hash")
	}
	return nil
}
